plugins/Compressor/Compressor.cpp

```diff
@@ -0,0 +1,651 @@
+				break;
+			}
+			case 2:// Average
+			{
+				gainResult[0] = gainResult[1] = (gainResult[0] + gainResult[1]) * 0.5f;
+				break;
+			}
+			case 3:// Minimum
+			{
+				gainResult[0] = gainResult[1] = qMax(gainResult[0], gainResult[1]);
+				break;
+			}
+			case 4:// Blend
+			{
+				if (blend > 0)// 0 is unlinked
+				{
+					if (blend <= 1)// Blend to minimum volume
+					{
+						const float temp1 = qMin(gainResult[0], gainResult[1]);
+						gainResult[0] = linearInterpolate(gainResult[0], temp1, blend);
+						gainResult[1] = linearInterpolate(gainResult[1], temp1, blend);
+					}
+					else if (blend <= 2)// Blend to average volume
+					{
+						const float temp1 = qMin(gainResult[0], gainResult[1]);
+						const float temp2 = (gainResult[0] + gainResult[1]) * 0.5f;
+						gainResult[0] = linearInterpolate(temp1, temp2, blend - 1);
+						gainResult[1] = gainResult[0];
+					}
+					else// Blend to maximum volume
+					{
+						const float temp1 = (gainResult[0] + gainResult[1]) * 0.5f;
+						const float temp2 = qMax(gainResult[0], gainResult[1]);
+						gainResult[0] = linearInterpolate(temp1, temp2, blend - 2);
+						gainResult[1] = gainResult[0];
+					}
+				}
+				break;
+			}
+		}
+
+		// Bias compression to the left or right (or mid or side)
+		if (stereoBalance != 0)
+		{
+			gainResult[0] = 1 - ((1 - gainResult[0]) * (stereoBalance > 0 ? 1 - stereoBalance : 1));
+			gainResult[1] = 1 - ((1 - gainResult[1]) * (stereoBalance < 0 ? 1 + stereoBalance : 1));
+		}
+
+		// For visualizer
+		m_displayGain[0] = gainResult[0];
+		m_displayGain[1] = gainResult[1];
+
+		// Delay the signal by 20 ms via ring buffer if lookahead is enabled
+		if (lookahead)
+		{
+			++m_inputBufLoc;
+			if (m_inputBufLoc >= m_lookaheadDelayLength)
+			{
+				m_inputBufLoc = 0;
+			}
+
+			const float temp[2] = {drySignal[0], drySignal[1]};
+			s[0] = m_inputBuf[0][m_inputBufLoc];
+			s[1] = m_inputBuf[1][m_inputBufLoc];
+
+			m_inputBuf[0][m_inputBufLoc] = temp[0];
+			m_inputBuf[1][m_inputBufLoc] = temp[1];
+		}
+		else
+		{
+			s[0] = drySignal[0];
+			s[1] = drySignal[1];
+		}
+
+		float delayedDrySignal[2] = {s[0], s[1]};
+		
+		if (midside)// Convert left/right to mid/side
+		{
+			const float temp = s[0];
+			s[0] = (temp + s[1]) * 0.5;
+			s[1] = temp - s[1];
+		}
+
+		s[0] *= gainResult[0] * m_outGainVal * (outBalance > 0 ? 1 - outBalance : 1);
+		s[1] *= gainResult[1] * m_outGainVal * (outBalance < 0 ? 1 + outBalance : 1);
+
+		if (midside)// Convert mid/side back to left/right
+		{
+			const float temp1 = s[0];
+			const float temp2 = s[1] * 0.5;
+			s[0] = temp1 + temp2;
+			s[1] = temp1 - temp2;
+		}
+
+		m_prevOut[0] = s[0];
+		m_prevOut[1] = s[1];
+
+		// Negate wet signal from dry signal
+		if (audition)
+		{
+			s[0] = -s[0] + delayedDrySignal[0];
+			s[1] = -s[1] + delayedDrySignal[1];
+		}
+		else if (autoMakeup)
+		{
+			s[0] *= m_autoMakeupVal;
+			s[1] *= m_autoMakeupVal;
+		}
+
+		// Calculate wet/dry value results
```

Comment:
As a reminder: proper latency compensation between W/D signal is required when lookahead is enabled. It's even possible to provide both fixed and minimum latency mode, but that's not that important until we implement general latency compensation feature.

Reply:
- If you use `delayedDrySignal[0]`, the dry signal will be affected by the input gain
- If you use `delayedDrySignal[0] / m_inGainVal`, you'll get weird gain artifacts on automating the input gain
- if you use `buf[f][0]`, you will get an artifact like flanger when lookahead is enabled

The solution is storing the dry signal **before** applying input gain into a ring buffer(`delayedDrySignal` is the signal after applying input gain). I'm not sure if the values for other controls should be evaluated before or after the delay.